common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java

```diff
@@ -0,0 +1,95 @@
+                return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL, error, errorDescription);
+
+            default:
+                return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                        MicrosoftAuthorizationErrorResponse.UNKNOWN_ERROR, MicrosoftAuthorizationErrorResponse.UNKNOWN_RESULT_CODE);
+        }
+    }
+
+    private MicrosoftStsAuthorizationResult createAuthorizationResultWithErrorResponse(final AuthorizationStatus authStatus,
+                                                                                       final String error,
+                                                                                       final String errorDescription) {
+        Logger.info(TAG, null, "Error is returned from webview redirect");
+        Logger.infoPII(TAG, null, "error: " + error + " errorDescription: " + errorDescription);
+        AuthorizationErrorResponse errorResponse = new MicrosoftStsAuthorizationErrorResponse(error, errorDescription);
+        return new MicrosoftStsAuthorizationResult(authStatus, errorResponse);
+    }
+
+    private MicrosoftStsAuthorizationResult parseUrlAndCreateAuthorizationResponse(final String url) {
+        HashMap<String, String> urlParameters = StringUtil.isEmpty(url) ? null : StringExtensions.getUrlParameters(url);
+        if (urlParameters == null || urlParameters.isEmpty()) {
+            Logger.warn(TAG, null, "Invalid server response, empty query string from the webview redirect.");
+            return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                    MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+        } else if (urlParameters.containsKey(CODE)) {
+            return validateAndCreateAuthorizationResult(urlParameters.get(CODE), urlParameters.get(STATE));
+        } else if (urlParameters.containsKey(ERROR)) {
+            return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    urlParameters.get(ERROR), urlParameters.get(ERROR_DESCRIPTION));
+        }
+        return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                MicrosoftAuthorizationErrorResponse.AUTHORIZATION_FAILED,
+                MicrosoftAuthorizationErrorResponse.AUTHORIZATION_SERVER_INVALID_RESPONSE);
+
+    }
+
+    private MicrosoftStsAuthorizationResult validateAndCreateAuthorizationResult(final String code, final String state) {
+        if (StringUtil.isEmpty(state)) {
+            Logger.warn(TAG, "State parameter is not returned from the webview redirect.");
+            return createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
+                    ErrorStrings.STATE_MISMATCH,
+                    MicrosoftAuthorizationErrorResponse.STATE_NOT_RETURNED);
+        } else {
+            //TODO : validate state
```

Comment:
Not requesting a change, but some things to think about for the future:
- How might this method signature change once state validation is included?
- Will is throw an Exception? Checked or unchecked?
    * Who will catch this Exception and what can that class do to propagate the Exception back to the caller?
- Which class will be responsible for performing state validation? Will this logic be reusable?

Reply:
Correct me if I am wrong but per my understanding we wouldn't be throwing an exception here. We would construct an `AuthorizationErrorResponse` with `MicrosoftAuthorizationErrorResponse.STATE_NOT_THE_SAME` and return back the `AuthorizationResult`. The calling class  should handle this error like any other result having `AuthorizationErrorMessage` and do the needful. 

As for which class would be responsible for state validation, I'm thinking of a resuable util method/class which compares the state from request and response. But this should not effect the structure of this class.

Reply:
Right on -- sounds good to me 👍